src/ucp/core/ucp_worker.h

```diff
@@ -25,25 +25,44 @@
 #define UCP_WORKER_HEADROOM_PRIV_SIZE 24
 
 
+#if ENABLE_MT
+
 #define UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(_worker)                 \
-    do {                                                                   \
-        ucs_assert(!UCP_THREAD_IS_REQUIRED(&(_worker)->mt_lock) ||      \
-                   UCP_THREAD_CS_IS_RECURSIVELY_LOCKED(&(_worker)->mt_lock) || \
-                   !UCS_ASYNC_IS_RECURSIVELY_BLOCKED(&(_worker)->async)); \
-        UCP_THREAD_CS_ENTER_CONDITIONAL(&(_worker)->mt_lock);           \
+    do {                                                                \
+        if ((_worker)->flags & UCP_WORKER_FLAG_MT) {                    \
```

Comment:
looks like there are lots of branches here, need to optimize (1-2 branches should be executed at most)

Reply:
Do you mean `if...else if..` implementation of `UCS_ASYNC_BLOCK`? On default spin lock path, there are the same number of conditions:
1) `flags & UCP_WORKER_FLAG_MT` instead of `ucp_mt_lock_t.mt_type != 0`
2) `((_async)->mode == UCS_ASYNC_MODE_THREAD_SPINLOCK)` instead of `((_lock_ptr)->mt_type == UCP_MT_TYPE_MUTEX)`

(2) may be replaced by indirect call but i don't think it would be more optimal.